### PR TITLE
Fix 23175 - -preview=in silently adds possible stack memory escape

### PR DESCRIPTION
**Pull request by dkorpel**

### CODE REVIEW

**Comment by maxhaton** [comment]
@Geod24

**Comment by thewilsonator** [comment]
Should this target stable?

**Review by Geod24 — CHANGES_REQUESTED** [review]
I commented on the issue itself which is IMO invalid.

First, the very idea that code behaving differently under a `-preview` switch is a breaking change is something we should reject instantly, as the whole point of a `-preview` switch is to allow developers to make such changes.

Second, that issue essentially says: "When I use `-preview=in`, a switch meant to make `in` behave as `const scope [ref]`, suddenly my parameters previously marked `in` behave as `scope`. Please fix."

**Comment by kinke** [comment]
This would break a perfectly valid:
```
@nogc:
void foo(in string s) {}
void main() { foo(['a']); }
```

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -2052,8 +2052,14 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
...
             const explicitScope = p.isLazy() ||
                 ((p.storageClass & STC.scope_) && !(p.storageClass & STC.scopeinferred));
+
+            // https://issues.dlang.org/show_bug.cgi?id=23175
+            // Don't trust `scope` from `in` while `-preview=in` transition is in progress
+            const scopeFromIn = global.params.previewIn && !(p.storageClass & STC.scope_) && (p.storageClass & STC.in_);
+
             if ((pStc & (STC.scope_ | STC.lazy_)) &&
                 ((global.params.useDIP1000 == FeatureState.enabled) || explicitScope) &&
+                !scopeFromIn &&
```
Comment (kinke) [inline src/dmd/expressionsem.d:2062]: Isn't the actual problem here that the DIP1000 check is too liberal and fails to detect this scope violation without `@safe`?

**Comment by schveiguy** [comment]
As I commented in the bug report, the current situation is making a breaking change on existing *valid* code without a warning to give the user a heads-up that this might happen. It's especially egregious because it's introducing memory corruption, which we all know is the worst form of bug to silently introduce.

So the *issue* is valid.

Now, regarding this PR, I'm not sure I agree with it. I prefer a warning to the user that this may happen in the case of a parameter that is marked `in`, but not `return`, yet is returned. I don't know how to make this happen, or if it's even feasible. But the feature itself, along with the optimization, should be allowed. Just warn when this is happening (maybe you can silence the warning with another switch?)

**Comment by Geod24** [comment]
> I prefer a warning to the user that this may happen in the case of a parameter that is marked `in`, but not `return`, yet is returned.

But why would we do this for `in`, but not for `scope` ?

**Comment by dkorpel** [comment]
I've now changed it into a transition flag.

**Comment by schveiguy** [comment]
> But why would we do this for `in`, but not for `scope` ?

Because `scope` is defined as "may not escape". `in` is defined as `const`. As of today, an `in` parameter may escape, and the compiler does not optimize away allocations based on that. With `scope`, it was *feasible* the compiler might do this, even if it didn't, so it's on you for relying on implementation details. With `in`, it wasn't legal for it do do this, until you changed the definition of `in`.

In other words, the feature turns well-defined code into undefined behavior, without any warning.

**Comment by Geod24** [comment]
> `in` is defined as `const`. 

Except for a brief period (< 5 months) in 2018, `in` has **always** be defined as `const scope`, as @pbackus [pointed out](https://issues.dlang.org/show_bug.cgi?id=23175#c3).

**Comment by schveiguy** [comment]
As I pointed out it has been const since feb 2021.

**Comment by Geod24** [comment]
@schveiguy : No, you said that before I changed the meaning of `in`,  it wasn't legal for it do do this (this being for the compiler to consider the parameter may not escape). That's plain wrong.

@dkorpel : I don't think that PR adds value, because the basis for the issue is incorrect. The issue triggers with `scope` as well. The change you've made to detect those trivial cases is IMO good, although I'm not sure everyone would agree ;)
If I'm not mistaken, said change will catch this case, correct ? Then we have `-transition=in` to list all instances of `in`, which IMO is better as the `ref` is likely to be more "surprising" to the user than `scope`.

**Comment by schveiguy** [comment]
In means const. Is it legal to make this optimization with const instead of in? If not then it’s not legal for in. “ When not in use, in is equivalent to const.”

**Comment by dkorpel** [comment]
> because the basis for the issue is incorrect. The issue triggers with `scope` as well. 

I don't have skin in this game, so I wouldn't mind if the issue was closed as WONTFIX. However, I do think @schveiguy has a point: it's reasonable to think legacy code has been written that assumes `in` parameters may escape the function. It's less reasonable to think this of `scope`, since the attribute's sole meaning is 'does not escape'. Yes, a preview switch means breaking changes are allowed, but usually that entails introducing new errors like `-preview=nosharedaccess`, or new edge case behavior like `-preview=intpromote`. Possibly introducing memory corruption without warning is a bit more severe than that, so a transition helper is in its place.

> The change you've made to detect those trivial cases is IMO good, although I'm not sure everyone would agree ;)

The change: https://github.com/dlang/dmd/pull/14221
I haven't heard any disagreement about this idea on the news group, what do you think the objection is?

> If I'm not mistaken, said change will catch this case, correct ? 

When 'this case' is `string foo(in string s) {return s;}` then yes. I had to change the test case in the test suite.

> Then we have `-transition=in` to list all instances of `in`, which IMO is better as the `ref` is likely to be more "surprising" to the user than `scope`.

I thought adding `ref` was supposed to silently work without problems.

**Comment by Geod24** [comment]
@schveiguy 
> In means const. Is it legal to make this optimization with const instead of in? If not then it’s not legal for in. “ **When not in use**, in is equivalent to const.”

So does the problem manifest when `-preview=in` **is not in use** ?

>  it's reasonable to think legacy code has been written that assumes `in` parameters may escape the function

That's where we disagree I think. `in` has been defined as `const scope` since 2013. I don't think it was ever reasonable to assume it could escape.

> I haven't heard any disagreement about this idea on the news group, what do you think the objection is?

Walter has been consistently opposed to the idea of bringing `scope` diagnostic to `@system` functions.

> I thought adding `ref` was supposed to silently work without problems.

I'm not sure if you are being ironic or not... :)
What I wanted, and the way it currently works, is that `in` (with `-preview=in`) means `const scope ref`, but sometimes decay to `const scope` if the compiler thinks that it's more efficient to pass by value (e.g. `int`, `bool`, etc...).
While some people agreed with this approach (@kinke for example), some people were vocally opposed to it (Andrei).
So the consensus we reached in one of the DLF meetings was to disable the optimization, so it would always mean `ref`. I'm going to implement it soon (as well as start preparation to enable `-preview=in` by default) as soon as #12242 is merged.

**Comment by dkorpel** [comment]
> Walter has been consistently opposed to the idea of bringing `scope` diagnostic to `@system` functions.

One of his latest comments suggested otherwise:
https://forum.dlang.org/post/t88iuv$mbb$1@digitalmars.com
 
> I'm not sure if you are being ironic or not... :) 

I'm not, I'm trying to recall what the problem with adding `ref` was. I think it was mutable aliasing:
```D
void f(in string x, ref string y) 
{
    const copy = x;
    y = "change";
    assert(x == copy); // when calling `f(s, s)`, succeeds or fails depending on whether `in` means `ref`
}
```

**Comment by Geod24** [comment]
> One of his latest comments suggested otherwise:
> https://forum.dlang.org/post/t88iuv$mbb$1@digitalmars.com

Haven't seen that one, great :)

> I'm not, I'm trying to recall what the problem with adding `ref` was. I think it was mutable aliasing: [...]

Yes it was. A problem that exists currently. The main point being that one couldn't "rely" on `in` being `ref`, so we might see "different behavior". I pointed out that this problem also exists for `auto ref`, that `@live` is supposed to fix this (how I don't know, but that was the promise ;) ) and that this is just bad code hat would break anyway (logically break, but not corrupt memory).

Nevertheless, I'd rather loose the optimization that the whole of `-preview=in`, so I'll disable it for DMD.

**Comment by schveiguy** [comment]
> @schveiguy
> 
> In means const. Is it legal to make this optimization with const instead of in? If not then it’s not legal for in. “ When not in use, in is equivalent to const.”
> 
> So does the problem manifest when -preview=in is not in use ?
> 

No because without the preview switches, the behavior is defined. In is currently const, as the quote says. With preview in it becomes scope const. If there are cases such as this one where preview in turns pre existing valid code into undefined behavior, we need to warn on that behavior. It doesn’t mean we have to disable the optimization, we just need to make sure the user is aware of what will happen if they don’t remedy the situation. 
> it's reasonable to think legacy code has been written that assumes in parameters may escape the function
> 
> That's where we disagree I think. in has been defined as const scope since 2013. I don't think it was ever reasonable to assume it could escape.
> 

This is not true. It’s also not an opinion, it’s a fact. In is defined as const since at least February 2021 in the spec, much longer in the implementation, IIUC.

**Comment by dkorpel** [comment]
Closing this because this is stalled. I think it's better to pursue Paul Backus' suggestion:

> The most straightforward fix would be to have the compiler emit a warning any time an `in` parameter escapes from a function in `@system` or `@trusted` code (i.e., the same thing it currently emits an error for in `@safe` code).

https://issues.dlang.org/show_bug.cgi?id=23175#c12
